Refactored RecepientScene & ViewModel

gemdev111 opened the pull request:
Moved model-related logic from scene to viewmodel 
 Added text validator protocol 
 Integrated manual validation and disabling logic for resolver validation and input validation to vm 
 Minor cleanup
 Added tests

gemdev111 committed:
Refactored RecepientScene & ViewModel
Moved model-related logic from scene to viewmodel
Added text validator protocol
Inegrated manual validation and disabling logic for resolver validation and input validation to vm
Minor cleanup

gemdev111 committed:
Minor updates
Added tests for InputValidationFieldViewModel

gemdev111 committed:
Merge branch 'main' into fields-error-handling

gemdev111 committed:
Minor naming updates

DRadmir commented on Gem/Transfer/Types/AddressValidator.swift:
```diff
@@ -0,0 +1,27 @@
+import Components
+import Primitives
+import Keystore
+import PrimitivesComponents
```
Could you please check, if all these imports are needed?

DRadmir commented on Gem/Transfer/Types/AddressValidator.swift:
Is this the right place for this file?

gemdev111 replied:
yeah, in future might be separate module related to transfer

DRadmir commented on Gem/Transfer/ViewsModels/RecipientSceneViewModel.swift:
```diff
@@ -8,54 +8,96 @@ import PrimitivesComponents
 }
 
-enum RecipientScanResult {
+private enum RecipientScanResult {
```
Is it possible to move these enums to the Types folder?

DRadmir commented on Gem/Transfer/ViewsModels/RecipientSceneViewModel.swift:
```diff
@@ -67,33 +109,97 @@ class RecipientViewModel: ObservableObject {
+
+    func onSelectPaste(field: RecipientScene.Field) {
+        guard let string = UIPasteboard.general.string else { return }
+        switch field {
```
All cases of this enum do the same, is it possible to remove the switch?

DRadmir replied:
Should memo be assigned something like `memo = string`?

DRadmir commented on Packages/PrimitivesComponents/Sources/Components/InputValidationField.swift:
```diff
@@ -0,0 +1,49 @@
+
+    var style: FloatFieldStyle
+    var placeholder: String
+    var allowClean: Bool
```
Should these properties be var?

DRadmir commented on Packages/PrimitivesComponents/Sources/Types/TextValidator.swift:
```diff
@@ -0,0 +1,11 @@
+}
+
+public struct UnknownValidationError: LocalizedError {
+    public var errorDescription: String? { "error.unknown" }
```
Could you consider using Localized instead of "error.unknown"?

DRadmir commented on Packages/PrimitivesComponents/Sources/ViewModels/InputValidationViewModel.swift:
```diff
@@ -0,0 +1,90 @@
+
+    public private(set) var error: (any LocalizedError)?
+
+    private var mode: InputValidationMode
```
```suggestion
    private let mode: InputValidationMode
```

DRadmir commented on Packages/PrimitivesComponents/Sources/ViewModels/InputValidationViewModel.swift:
```diff
@@ -0,0 +1,90 @@
+        return validate()
+    }
+
+    public func update(customError: (any Error)?) {
```
```suggestion
    public func update(error: (any Error)?) {
```

DRadmir commented on Packages/PrimitivesComponents/Sources/ViewModels/InputValidationViewModel.swift:
```diff
@@ -0,0 +1,90 @@
+        return error == nil
+    }
+
+    public func updateValidators(_ validators: [any TextValidator]) {
```
For consistency with other methods
```suggestion
    public func update(validators: [any TextValidator]) {
```

DRadmir commented on Packages/PrimitivesComponents/Sources/ViewModels/InputValidationViewModel.swift:
```diff
@@ -0,0 +1,90 @@
+        _ text: String,
+        with validators: [any TextValidator]
+    ) -> (any LocalizedError)? {
+        guard !text.isEmpty else { return nil }
```
```suggestion
        guard text.isNotEmpty else { return nil }
```

gemdev111 committed:
minor improvements

gemdev111 committed:
removed unused improt

gemcoder21 commented on Gem/Transfer/ViewsModels/RecipientSceneViewModel.swift:
```diff
@@ -8,54 +8,85 @@ import PrimitivesComponents
+
+        self.addressInputModel = InputValidationViewModel(
+            mode: .manual,
+            validators: [AddressValidator(asset: asset)]
```
```suggestion
            validators: [AddressValidator(asset: asset)]
```

can you add extension so we can setup as `validators: [.required, .addressValidation(Chain), .minAmount(1)]` lomda ;ole tjos

gemcoder21 commented on Gem/Transfer/ViewsModels/RecipientSceneViewModel.swift:
```diff
@@ -67,33 +98,94 @@ class RecipientViewModel: ObservableObject {
+            recipient: Recipient(
+                name: recipient.name,
+                address: recipient.address,
+                memo: nil // TODO: - add memo later
```
should we add memo?

gemcoder21 commented on Gem/Transfer/ViewsModels/RecipientSceneViewModel.swift:
```diff
@@ -67,33 +98,94 @@ class RecipientViewModel: ObservableObject {
+
     //TODO: Add unit tests
-    func getPaymentScanResult(string: String) throws -> PaymentScanResult {
+    private func decodeScanned(string: String) throws -> PaymentScanResult {
```
```suggestion
    private func paymentScan(for string: String) throws -> PaymentScanResult {
```

gemcoder21 commented on Gem/Transfer/ViewsModels/RecipientSceneViewModel.swift:
```diff
@@ -102,18 +194,16 @@ class RecipientViewModel: ObservableObject {
-            (showMemo ? ((payment.memo?.isEmpty) == nil) : true),
-            asset.chain.isValidAddress(payment.address)
+
+    private func toRecipientScanResult(payment: PaymentScanResult) throws -> RecipientScanResult {
```
```suggestion
    private func getRecipientScanResult(payment: PaymentScanResult) throws -> RecipientScanResult {
```

gemcoder21 commented on Gem/Transfer/ViewsModels/RecipientSceneViewModel.swift:
```diff
@@ -67,33 +98,94 @@ class RecipientViewModel: ObservableObject {
     }
-    
+
     //TODO: Add unit tests
```
can you add unit test?

gemcoder21 commented on Gem/Transfer/ViewsModels/RecipientSceneViewModel.swift:
```diff
@@ -160,47 +242,52 @@ class RecipientViewModel: ObservableObject {
+            let scanResult = try toRecipientScanResult(payment: payment)
+            switch scanResult {
+            case .transferData(let data):
+                handle(data: data)
```
```suggestion
                handleTransferData(data)
```

gemcoder21 commented on Packages/PrimitivesComponents/Sources/Components/InputValidationField.swift:
```diff
@@ -0,0 +1,49 @@
+                }
+            )
+
+            if let message = model.error?.errorDescription {
```
can we animate when showing an error? in 0.1 sec

gemcoder21 commented on Packages/PrimitivesComponents/Sources/Types/TextValidator.swift:
```diff
@@ -0,0 +1,12 @@
+    func validate(_ text: String) throws
+}
+
+public struct UnknownValidationError: LocalizedError {
```
Is it possible to remove `UnknownValidationError`

gemcoder21 commented on Packages/PrimitivesComponents/Sources/ViewModels/InputValidationViewModel.swift:
```diff
@@ -0,0 +1,90 @@
+        self.validators = validators
+    }
+
+    public var isValid: Bool { error == nil && text.isNotEmpty }
```
`text.isNotEmpty` this might not work if you have validator for required field?

gemcoder21 commented on Packages/PrimitivesComponents/Sources/ViewModels/InputValidationViewModel.swift:
```diff
@@ -0,0 +1,90 @@
+extension InputValidationViewModel {
+
+    @discardableResult
+    public func validate() -> Bool {
```
maybe call it 
```
func isValid() -> Error? {
   Self.validate(text, with: validators)
}
```
and set aror in another place

gemcoder21 commented on Packages/PrimitivesComponents/Tests/PrimitivesComponentsTests/InputValidationViewModelTests.swift:
```diff
@@ -0,0 +1,166 @@
+    case boom
+}
+
+private struct StubValidator: TextValidator {
```
can we remove this one `StubValidator`? let's test using real validators

gemcoder21 commented on Packages/PrimitivesComponents/Tests/PrimitivesComponentsTests/InputValidationViewModelTests.swift:
```diff
@@ -0,0 +1,166 @@
+    @Test
+    func combinedValidatorsFailWhenStubFails() {
+        let vm = InputValidationViewModel(
+            validators: [
```
you should test each individual validator instead

gemdev111 committed:
Updates
updates RecipientSceneViewModel
improved tests, InputValidationViewModel
moved validators logic to primitives-components
added tests for validators
minor refactoring
added localizations

gemdev111 committed:
Update core

gemdev111 committed:
Merge branch 'main' into fields-error-handling

gemdev111 committed:
removed unused @Environment

gemdev111 committed:
moved RecipientScanResult to types

gemdev111 committed:
Merge branch 'main' into fields-error-handling

DRadmir commented on Gem/Transfer/Navigation/RecipientNavigationView.swift:
```diff
@@ -3,69 +3,50 @@
             AmountNavigationView(
-                input: AmountInput(type: .transfer(recipient: data), asset: asset),
-                wallet: wallet,
+                input: AmountInput(type: .transfer(recipient: data), asset: model.asset),
```
Consider returning AmountInput from the model

DRadmir commented on Packages/PrimitivesComponents/Sources/Errors/RequiredFieldError.swift:
```diff
@@ -0,0 +1,17 @@
+import Foundation
+import Localization
+
+public enum RequiredFieldError: Equatable {
```
what other cases could there be? Perhaps it's better to use a struct?

DRadmir commented on Packages/PrimitivesComponents/Sources/Errors/RequiredFieldError.swift:
```diff
@@ -0,0 +1,17 @@
+
+extension RequiredFieldError: LocalizedError {
+    public var errorDescription: String? {
+        guard case .field(let name) = self else {
```
If we decide to use an enum, I suggest using a switch statement. This way, when a new case is added, the developer won't forget to handle it

DRadmir commented on Packages/PrimitivesComponents/Sources/Validators/AddressValidator.swift:
```diff
@@ -0,0 +1,27 @@
+public struct AddressValidator: TextValidator {
+    private let asset: Asset
+
+    init(asset: Asset) {
```
```suggestion
    public init(asset: Asset) {
```

DRadmir commented on Packages/PrimitivesComponents/Sources/Validators/AddressValidator.swift:
```diff
@@ -0,0 +1,27 @@
+}
+
+public extension TextValidator where Self == AddressValidator {
+    static func address(_ asset: Asset) -> AddressValidator {
```
The method name is a bit confusing. Maybe better something like addressValidator would be better?

gemdev111 replied:
```
validators: [
                .required(requireName: Localized.Transfer.Recipient.addressField),
                .address(asset)
            ]
            ```
            That is Tim suggestion with naming, and I also like it. You still can do AddressValidator(...) and also 
            .address is shorthand for this. It's clear when you put in in validators array

DRadmir commented on Packages/PrimitivesComponents/Sources/Validators/RequiredValidator.swift:
```diff
@@ -0,0 +1,25 @@
+}
+
+public extension TextValidator where Self == RequiredValidator {
+    static func required(requireName: String) -> RequiredValidator {
```
```suggestion
    static func requiredValidator(_ name: String) -> RequiredValidator {
```

gemdev111 replied:
same here , that is shorthand variation

DRadmir commented on Packages/WalletCorePrimitives/Sources/WalletCorePrimitives/Chain+WalletCorePrimitives.swift:
```diff
@@ -55,4 +55,8 @@ public extension Chain {
     }
+    
+    func isValidAddress(_ address: String) -> Bool {
+        return AnyAddress.isValid(string: address, coin: coinType)
```
```suggestion
        AnyAddress.isValid(string: address, coin: coinType)
```

DRadmir committed:
Merge branch 'main' into fields-error-handling

gemdev111 committed:
Updates

gemdev111 committed:
Update core

gemdev111 committed:
improved addressvalidator init

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
fix tests after adjustements
